### DISCUSSION TITLE
Replace current page content with 404 page without changing URL

### DISCUSSION THREAD

**Discussion by nicholaschiang** [vercel/next.js#21154] [Ideas]
**Describe the feature you'd like to request**

Right now, I have automatically statically optimized pages like an admin dashboard that only certain users have access to. To control that access, I have an effect that:
1. Checks if the user is logged in. Upon initial page load, the `loggedIn` state will be `undefined` and will change to `false` or `true` once the `/api/account` endpoint is requested.
2. If the user is logged in **and** we're still getting 401 errors when loading the page data, I redirect to the 404 page.
3. If the user is logged out, I redirect to the login page.

```typescript
import Router from 'next/router';
...
const { loggedIn } = useUser();
const { data, error } = useSWR('/api/admin-dashboard-data');
useEffect(() => {
  if (loggedIn && error?.code === 401) {
    void Router.replace('/404');
  }
}, [loggedIn, error]);
```

**Describe the solution you'd like**

Instead of just `Router.replace` the current page with the 404 page, it'd be nice if Next.js provided an easy way to keep the current URL but display the 404 page instead of the current page. This would be identical to how Next.js behaves when [returning a `notFound` boolean in `getStaticProps`](https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation).

**Describe alternatives you've considered**

Creating an `_app.tsx` level context myself that can conditionally show the 404 page instead of the current page:
```tsx
// _app.tsx
export default function App({ Component, pageProps }: AppProps): JSX.Element {
  const [errored, setErrored] = useState<boolean>(false);
  return (
    <ErrorContext.Provider value={{ setErrored }}>
      {!errored && <Component {...pageProps} />}
      {errored && <Custom404Page />}
    </ErrorContext.Provider>
  );
}
```
And then my page would then consume that `ErrorContext` and updating it would show the 404 page instead of the current page.

**Comment by lalit2005** [comment]
Can `rewrites` solve the problem?

**Reply by nicholaschiang** [reply]
It has to be dynamic and determined client-side after data fetching with SWR fails.

**Comment by bymoe** [comment]
I had this problem and `notFound` solved it for me
https://nextjs.org/docs/api-reference/data-fetching/get-server-side-props#notfound

**Reply by GabenGar** [reply]
This is only for server-rendered pages. In his case he has to create a [custom error page](https://nextjs.org/docs/advanced-features/custom-error-page) and show it on client when needed.

**Comment by ConnorMooneyhan** [comment]
Was a solution ever found for this problem? I'm having the same one and have tried a lot without getting anywhere.

**Reply by GabenGar** [reply]
What exactly did you try? Pretty sure the solution is to have 404 page as a component which is used by `404` page always and conditionally outside of it.
